spatialdb/utils/add_lookup_keys_to_s3_index.py

```diff
@@ -0,0 +1,253 @@
+            worker_num (optional[int]): Zero-based worker number if parallelizing.
+            num_workers (optional[int]): Total number of workers.
+        """
+        self.dynamodb = boto3.client('dynamodb', region_name=region)
+        self.table = table_name
+        self.max_items = max_items
+        self.worker_num = worker_num
+        self.num_workers = num_workers
+        if worker_num >= num_workers:
+            raise ValueError('worker_num must be less than num_workers')
+
+    def start(self):
+        """
+        Starts scan and update of S3 index table.
+        """
+        exclusive_start_key = None
+        done = False
+        count = 0
+        while not done:
+            resp = self.scan(exclusive_start_key)
+            if resp is None:
+                continue
+
+            if 'LastEvaluatedKey' not in resp or len(resp['LastEvaluatedKey']) == 0:
+                done = True
+            else:
+                exclusive_start_key = resp['LastEvaluatedKey']
+
+            print('Consumed read capacity: {}'.format(resp['ConsumedCapacity']))
+
+            for item in resp['Items']:
+                print(item)
+                self.add_lookup_key(item)
+
+            if exclusive_start_key is not None:
+                print('Continuing scan following {} - {}.'.format(
+                    exclusive_start_key[OBJ_KEY]['S'], 
+                    exclusive_start_key[VERSION_NODE]['N']))
+
+            count+=1
+
+            # Terminate early for testing.
+            #if count > 2:
+            #    done = True
+
+        print('Update complete.')
+
+    def scan(self, exclusive_start_key=None):
+        """
+        Invoke DynamoDB.Client.scan() and get up to self.max_items.
+
+        Will use exponential backoff and retry 4 times if throttled by Dynamo.
+
+        Args:
+            exclusive_start_key (dict): If defined, start scan from this point in the table.
+
+        Returns:
+            (dict|None): Response dictionary from DynamoDB.Client.scan() or None if throttled repeatedly.
+        """
+        scan_args = {
+            'TableName': self.table,
+            'Limit': self.max_items,
+            'ProjectionExpression': '#objkey,#vernode,#lookupkey',
+            'FilterExpression':'attribute_not_exists(#lookupkey)',
+            'ExpressionAttributeNames': {
+                '#lookupkey': LOOKUP_KEY,
+                '#objkey': OBJ_KEY,
+                '#vernode': VERSION_NODE
+            },
+            'ConsistentRead': True,
+            'ReturnConsumedCapacity': 'TOTAL'
+        }
+
+        if exclusive_start_key is not None:
+            scan_args['ExclusiveStartKey'] = exclusive_start_key
+
+        if self.num_workers > 1:
+            scan_args['Segment'] = self.worker_num
+            scan_args['TotalSegments'] = self.num_workers
+
+        for backoff in range(0, 6):
+            try:
+                resp = self.dynamodb.scan(**scan_args)
+                return resp
```

Comment:
should check for 200?

Reply:
I think boto3 throws an exception on a non-200 response so I've stopped checking for 200s. It seems like anytime anything goes wrong, there's an exception raised from what I've seen.